File: C++/54. Spiral Matrix.cpp
class Solution {
public:
    vector<int> spiralOrder(vector<vector<int>>& m) {        
        vector<int> q;
        int t = 0, d = m.size() - 1;
        int l = 0, r = m[0].size() - 1;
        while(t <= d && l <= r) {
            for(int j = l; j <= r; j++) {
                q.push_back(m[t][j]);
            }
            t++;
            for(int i = t; i <= d; i++) {
                q.push_back(m[i][r]);
            }
            r--;
            if(t <= d) {
                for(int j = r; j >= l; j--) {
                    q.push_back(m[d][j]);
                }
                d--;
            }
            if(l <= r) {
                for(int i = d; i >= t; i--) {
                    q.push_back(m[i][l]);
                }
                l++;
            }
        }
        return q;
    }
};

File: JAVA/54. Spiral Matrix.java
class Solution {
    public List<Integer> spiralOrder(int[][] m) {
        List<Integer> q = new ArrayList<Integer>();
        int i,t=0,l=0,d=m.length-1,r=m[0].length-1;
        while(t<=d && l<=r){
            for(i=l;i<=r;i++){
                q.add(m[t][i]);
            }
            t++;
            for(i=t;i<=d;i++){
                q.add(m[i][r]);
            }
            r--;
            if(t<=d){
            for(i=r;i>=l;i--){
                q.add(m[d][i]);
            }
            d--;}
            if(l<=r){
            for(i=d;i>=t;i--){
                q.add(m[i][l]);
            }}
            l++;;
        }
        return q;
    }
}